### PR TITLE
[IMPL] GUI Saving & Breadcrumbs

### PR DESCRIPTION
**Pull request by Wolfsurge**

### CODE REVIEW

**Comment by linustouchtips** [comment]
so awesome

**Comment by Wolfsurge** [comment]
ikr

File: src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import static org.lwjgl.opengl.GL11.*;
+
+/**
+ * @author Wolfsurge
+ * @since 01/04/22
+ */
+public class BreadcrumbsModule extends Module {
+
+    // Settings
+    public static final Setting<Boolean> infinite = new Setting<>("Infinite", false).setDescription("Makes breadcrumbs last forever");
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:21]: move settings and variables below the super call

File: src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    public static final Setting<Float> lifespanValue = new Setting<>("Lifespan", 10F, 100F, 1000F, 0).setDescription("The lifespan of the positions in ticks").setVisible(() -> !infinite.getValue());
+    public static final Setting<Float> lineWidth = new Setting<>("LineWidth", 0.1F, 3F, 5F, 1).setDescription("The width of the lines");
+
+    // List of positions
+    private final ArrayList<Position> positions = new ArrayList<>();
+
+    public BreadcrumbsModule() {
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:28]: don't forget to create an instance variable

File: src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        super("Breadcrumbs", Category.VISUAL, "Draws a trail behind you");
+    }
+
+    @Override
+    public void onEnable() {
+        // Clear positions
+        positions.clear();
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:35]: no need, you already cleared it when we disabled the module

File: src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    @Override
+    public void onDisable() {
+        // Clear positions
+        positions.clear();
+    }
+
+    @Override
+    public void onTick() {
+        if (!nullCheck() || mc.player.ticksExisted <= 20) {
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:46]: pretty sure there's an event for this, iirc it was like ConnectEvent or something

File: src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        glEnable(GL_TEXTURE_2D);
+        glPopMatrix();
+    }
+
+    static class Position {
+        // The position's position
+        private final Vec3d position;
+
+        // The position's lifespan
+        private long lifespan = lifespanValue.getValue().longValue();
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:102]: i think you could do this with a map and System.currentTimeMillis()

File: src/main/java/cope/cosmos/client/Cosmos.java [hunk 6]
```diff
@@ -147,6 +147,7 @@ public void init(FMLInitializationEvent event) {
 
         // sets up the GUI
         clickGUI = new ClickGUIScreen();
+        presetManager.loadGUI();
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/Cosmos.java:150]: add this to the existing load function

File: src/main/java/cope/cosmos/client/ui/clickgui/ClickGUIScreen.java [hunk 7]
```diff
@@ -181,6 +181,9 @@ public void onGuiClosed() {
         // save our configs when exiting the GUI
         Cosmos.INSTANCE.getPresetManager().save();
 
+        // Save GUI config
+        Cosmos.INSTANCE.getPresetManager().saveGUI();
```
Comment (linustouchtips) [inline src/main/java/cope/cosmos/client/ui/clickgui/ClickGUIScreen.java:185]: add this to the existing save function

File: src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java (hunk 5 above)
Comment (Wolfsurge) [reply src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:102]: can i just leave it i really cba

Comment (linustouchtips) [reply src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:102]: yea i can try to do it

Comment (Wolfsurge) [reply src/main/java/cope/cosmos/client/features/modules/visual/BreadcrumbsModule.java:102]: kk

**Pull request merged**